<a href="https://colab.research.google.com/github/anshudhanshu/Demand-Prediction-for-public-transport/blob/main/Sudhanshu_Rawat_Team_5_Capstone_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <b><u> Project Title : Traffic Jam: Predicting People's Movement into Nairobi </u></b>

## <b> Problem Description </b>

### This challenge asks you to build a model that predicts the number of seats that Mobiticket can expect to sell for each ride, i.e. for a specific route on a specific date and time. There are 14 routes in this dataset. All of the routes end in Nairobi and originate in towns to the North-West of Nairobi towards Lake Victoria.


### The towns from which these routes originate are:

* Awendo
* Homa Bay
* Kehancha
* Kendu Bay
* Keroka
* Keumbu
* Kijauri
* Kisii
* Mbita
* Migori
* Ndhiwa
* Nyachenge
* Oyugis
* Rodi
* Rongo
* Sirare
* Sori

### The routes from these 14 origins to the first stop in the outskirts of Nairobi takes approximately 8 to 9 hours from time of departure. From the first stop in the outskirts of Nairobi into the main bus terminal, where most passengers get off, in Central Business District, takes another 2 to 3 hours depending on traffic.

### The three stops that all these routes make in Nairobi (in order) are:

1. Kawangware: the first stop in the outskirts of Nairobi
2. Westlands
3. Afya Centre: the main bus terminal where most passengers disembark

### All of these points are mapped [here](https://www.google.com/maps/d/viewer?mid=1Ef2pFdP8keVHHid8bwju2raoRvjOGagN&ll=-0.8281897101491997%2C35.51706279999996&z=8).

### Passengers of these bus (or shuttle) rides are affected by Nairobi traffic not only during their ride into the city, but from there they must continue their journey to their final destination in Nairobi wherever that may be. Traffic can act as a deterrent for those who have the option to avoid buses that arrive in Nairobi during peak traffic hours. On the other hand, traffic may be an indication for people’s movement patterns, reflecting business hours, cultural events, political events, and holidays.

## <b> Data Description </b>

### <b>Nairobi Transport Data.csv (zipped)</b> is the dataset of tickets purchased from Mobiticket for the 14 routes from “up country” into Nairobi between 17 October 2017 and 20 April 2018. This dataset includes the variables: ride_id, seat_number, payment_method, payment_receipt, travel_date, travel_time, travel_from, travel_to, car_type, max_capacity.


### Uber Movement traffic data can be accessed [here](https://movement.uber.com). Data is available for Nairobi through June 2018. Uber Movement provided historic hourly travel time between any two points in Nairobi. Any tables that are extracted from the Uber Movement platform can be used in your model.

### Variables description:

* #### ride_id: unique ID of a vehicle on a specific route on a specific day and time.
* #### seat_number: seat assigned to ticket
* #### payment_method: method used by customer to purchase ticket from Mobiticket (cash or Mpesa)
* #### payment_receipt: unique id number for ticket purchased from Mobiticket
* #### travel_date: date of ride departure. (MM/DD/YYYY)
* #### travel_time: scheduled departure time of ride. Rides generally depart on time. (hh:mm)
* #### travel_from: town from which ride originated
* #### travel_to: destination of ride. All rides are to Nairobi.
* #### car_type: vehicle type (shuttle or bus)
* #### max_capacity: number of seats on the vehicle

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

In [3]:
import datetime

In [71]:
dataset = pd.read_csv('/content/drive/MyDrive/AlmaBetter/Cohort Aravali/Module 4/Week 1/Copy of Nairobi Transport Data.zip (Unzipped Files)/train_revised.csv')
dataset.head()

ride_id seat_number payment_method  ... travel_to car_type max_capacity
0     1442         15A          Mpesa  ...   Nairobi      Bus           49
1     5437         14A          Mpesa  ...   Nairobi      Bus           49
2     5710          8B          Mpesa  ...   Nairobi      Bus           49
3     5777         19A          Mpesa  ...   Nairobi      Bus           49
4     5778         11A          Mpesa  ...   Nairobi      Bus           49

[5 rows x 10 columns]

In [5]:
dataset.tail()

ride_id seat_number payment_method  ... travel_to car_type max_capacity
51640    13826          9B          Mpesa  ...   Nairobi      Bus           49
51641    13809         18A          Mpesa  ...   Nairobi      Bus           49
51642    13809         17A          Mpesa  ...   Nairobi      Bus           49
51643    13796         16B          Mpesa  ...   Nairobi      Bus           49
51644    14304           7          Mpesa  ...   Nairobi      Bus           49

[5 rows x 10 columns]

In [6]:
dataset['travel_date'].unique()

array(['17-10-17', '19-11-17', '26-11-17', '27-11-17', '20-04-18',
       '19-04-18', '18-04-18', '17-04-18', '16-04-18', '15-04-18',
       '14-04-18', '13-04-18', '12-04-18', '11-04-18', '28-11-17',
       '10-04-18', '09-04-18', '08-04-18', '07-04-18', '06-04-18',
       '05-04-18', '04-04-18', '03-04-18', '02-04-18', '01-04-18',
       '31-03-18', '30-03-18', '29-03-18', '28-03-18', '27-03-18',
       '29-11-17', '26-03-18', '25-03-18', '24-03-18', '23-03-18',
       '22-03-18', '21-03-18', '20-03-18', '19-03-18', '18-03-18',
       '17-03-18', '16-03-18', '30-11-17', '01-12-17', '15-03-18',
       '02-12-17', '14-03-18', '13-03-18', '03-12-17', '04-12-17',
       '12-03-18', '05-12-17', '11-03-18', '06-12-17', '10-03-18',
       '09-03-18', '07-12-17', '08-03-18', '08-12-17', '07-03-18',
       '06-03-18', '09-12-17', '10-12-17', '05-03-18', '04-03-18',
       '11-12-17', '12-12-17', '03-03-18', '02-03-18', '13-12-17',
       '01-03-18', '14-12-17', '28-02-18', '27-02-18', '15-12-

### **Understanding The data**

In [7]:
dataset.shape

(51645, 10)

There are 51645 observations in the dataset. And has 10 columns.

In [8]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51645 entries, 0 to 51644
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   ride_id          51645 non-null  int64 
 1   seat_number      51645 non-null  object
 2   payment_method   51645 non-null  object
 3   payment_receipt  51645 non-null  object
 4   travel_date      51645 non-null  object
 5   travel_time      51645 non-null  object
 6   travel_from      51645 non-null  object
 7   travel_to        51645 non-null  object
 8   car_type         51645 non-null  object
 9   max_capacity     51645 non-null  int64 
dtypes: int64(2), object(8)
memory usage: 3.9+ MB


In [9]:
dataset.describe(include='all')

ride_id seat_number  ... car_type  max_capacity
count   51645.000000       51645  ...    51645  51645.000000
unique           NaN          61  ...        2           NaN
top              NaN           1  ...      Bus           NaN
freq             NaN        2065  ...    31985           NaN
mean    10188.645793         NaN  ...      NaN     34.534321
std      2211.295708         NaN  ...      NaN     18.451193
min      1442.000000         NaN  ...      NaN     11.000000
25%      8287.000000         NaN  ...      NaN     11.000000
50%     10353.000000         NaN  ...      NaN     49.000000
75%     12029.000000         NaN  ...      NaN     49.000000
max     20117.000000         NaN  ...      NaN     49.000000

[11 rows x 10 columns]

In [10]:
dataset.isnull().sum()

ride_id            0
seat_number        0
payment_method     0
payment_receipt    0
travel_date        0
travel_time        0
travel_from        0
travel_to          0
car_type           0
max_capacity       0
dtype: int64

In [11]:
dataset.seat_number.unique()

array(['15A', '14A', '8B', '19A', '11A', '18B', '25', '21B', '22A', '17B',
       '1X', '3A', '1B', '1', '4A', '18A', '23A', '5', '16B', '8', '15B',
       '10', '3', '9', '7', '6', '4', '2', '20A', '9A', '22B', '23B',
       '24A', '2B', '13B', '24B', '14B', '20B', '19B', '10A', '17A', '2A',
       '7B', '6B', '5B', '3B', '4B', '1A', '5A', '6A', '21A', '13A', '8A',
       '9B', '12B', '10B', '7A', '16A', '12A', '11B', '14'], dtype=object)

In [12]:
dataset['travel_time'].unique()

array(['7:15', '7:12', '7:05', '7:10', '7:09', '5:10', '9:50', '7:06',
       '6:00', '5:50', '10:58', '6:50', '5:30', '6:40', '9:10', '5:00',
       '5:20', '10:52', '10:50', '10:51', '8:20', '19:03', '7:02', '6:30',
       '7:20', '7:04', '19:05', '19:04', '6:59', '7:08', '5:40', '9:00',
       '7:14', '7:07', '8:00', '7:30', '10:57', '8:30', '11:01', '7:11',
       '7:03', '6:58', '11:05', '10:20', '10:53', '10:00', '8:50',
       '10:54', '7:01', '7:00', '6:10', '6:20', '10:10', '11:02', '9:40',
       '9:30', '7:50', '10:56', '10:55', '10:40', '10:30', '9:20', '8:40',
       '8:10', '7:40', '7:13', '11:00', '11:04', '19:06', '10:59',
       '11:07', '11:06', '11:03', '19:07', '11:08', '11:09', '23:09',
       '23:10'], dtype=object)

In [13]:
dataset.travel_time.value_counts()

7:09     3926
7:04     3837
7:12     3540
19:07    3227
7:08     2989
         ... 
11:05      51
11:06      48
11:07      32
11:08      18
11:09       9
Name: travel_time, Length: 78, dtype: int64

In [72]:
dataset[['travel_date','seat_number']].groupby('travel_date').count().reset_index().sort_values(by='travel_date')

travel_date  seat_number
0      01-01-18           61
1      01-02-18          456
2      01-03-18          439
3      01-04-18          274
4      01-12-17          434
..          ...          ...
144    30-11-17           86
145    30-12-17          172
146    31-01-18          336
147    31-03-18          170
148    31-12-17          128

[149 rows x 2 columns]

In [73]:
traveler_per_day = pd.DataFrame(dataset[['travel_date','travel_from','travel_to']].groupby(['travel_date','travel_to','travel_from']).agg({'travel_from':'count'}))



In [74]:

traveler_per_day.rename(columns={'travel_from':'Sold_ticket'},inplace=True)

In [75]:
traveler_per_day = traveler_per_day.reset_index()

In [76]:

traveler_per_day['travel_date']

0       01-01-18
1       01-01-18
2       01-01-18
3       01-01-18
4       01-01-18
          ...   
1243    31-12-17
1244    31-12-17
1245    31-12-17
1246    31-12-17
1247    31-12-17
Name: travel_date, Length: 1248, dtype: object

In [77]:
traveler_per_day['travel_date'] = pd.to_datetime(traveler_per_day['travel_date'])

In [20]:
traveler_per_day.sort_values(by='travel_date',ascending=True)

travel_date travel_to travel_from  Sold_ticket
36   2017-01-12   Nairobi       Rongo           11
34   2017-01-12   Nairobi       Kisii          315
33   2017-01-12   Nairobi      Keumbu            1
32   2017-01-12   Nairobi      Keroka          106
35   2017-01-12   Nairobi      Migori            1
..          ...       ...         ...          ...
504  2018-12-04   Nairobi        Rodi           29
506  2018-12-04   Nairobi      Sirare           38
507  2018-12-04   Nairobi        Sori            5
495  2018-12-04   Nairobi    Homa Bay           70
505  2018-12-04   Nairobi       Rongo           42

[1248 rows x 4 columns]

In [21]:
traveler_per_day[['travel_date','Sold_ticket']].groupby('travel_date').sum()

Sold_ticket
travel_date             
2017-01-12           434
2017-02-12            71
2017-03-12           152
2017-04-12           326
2017-05-12           498
...                  ...
2018-11-04           391
2018-12-01           410
2018-12-02           508
2018-12-03           416
2018-12-04           410

[149 rows x 1 columns]

In [22]:
len(dataset['ride_id'].unique())

6249

In [23]:
dataset['ride_id'].value_counts()

8454     50
8090     49
10834    49
11382    49
13211    49
         ..
5804      1
9834      1
9738      1
11755     1
6147      1
Name: ride_id, Length: 6249, dtype: int64

In [24]:
column = dataset.columns

In [25]:
column

Index(['ride_id', 'seat_number', 'payment_method', 'payment_receipt',
       'travel_date', 'travel_time', 'travel_from', 'travel_to', 'car_type',
       'max_capacity'],
      dtype='object')

### Finding Our Target Variable

In [78]:
label = Target=dataset.groupby(["ride_id"]).seat_number.count().rename("number_of_ticket").reset_index()
label

ride_id  number_of_ticket
0        1442                 1
1        5437                 1
2        5710                 1
3        5777                 5
4        5778                31
...       ...               ...
6244    20113                 9
6245    20114                 4
6246    20115                 5
6247    20116                 1
6248    20117                 1

[6249 rows x 2 columns]

#### Now we found our target variable so let us delete the duplicate values from `ride_id` 

In [79]:
dataset=dataset.drop_duplicates("ride_id")

In [80]:
dataset.shape

(6249, 10)

In [81]:
dataset = dataset.merge(label, how="left",on="ride_id")
dataset.head()

ride_id seat_number payment_method  ... car_type max_capacity number_of_ticket
0     1442         15A          Mpesa  ...      Bus           49                1
1     5437         14A          Mpesa  ...      Bus           49                1
2     5710          8B          Mpesa  ...      Bus           49                1
3     5777         19A          Mpesa  ...      Bus           49                5
4     5778         11A          Mpesa  ...      Bus           49               31

[5 rows x 11 columns]

In [82]:
dataset = dataset.drop(['seat_number','payment_method','payment_receipt'], axis=1)

In [83]:
X = dataset.drop('number_of_ticket', axis=1)
y = dataset['number_of_ticket']

In [84]:
from sklearn.model_selection import train_test_split
from sklearn.feature_selection import VarianceThreshold

In [122]:
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.33, random_state=42)

### **Removing Constant Features**

We need to first remove the constant features. 

In [123]:
import warnings
warnings.filterwarnings('ignore')

In [124]:
X_train.drop(["travel_to"],axis=1,inplace=True)
X_test.drop(["travel_to"],axis=1,inplace=True)

In [126]:
X_train["date"]=X_train["travel_date"]+ " "+X_train["travel_time"]
X_test["date"]=X_test["travel_date"]+ " "+X_test["travel_time"]

In [127]:
X_train

ride_id travel_date travel_time  ... car_type max_capacity            date
4376    11559    12-12-17        6:00  ...  shuttle           11   12-12-17 6:00
690      6887    08-12-17        5:10  ...  shuttle           11   08-12-17 5:10
2653     9409    29-01-18        5:20  ...  shuttle           11   29-01-18 5:20
602      6784    09-03-18       10:10  ...  shuttle           11  09-03-18 10:10
5863    13338    11-04-18        8:20  ...  shuttle           11   11-04-18 8:20
...       ...         ...         ...  ...      ...          ...             ...
3772    10844    19-02-18        7:11  ...      Bus           49   19-02-18 7:11
5191    12499    20-03-18        5:00  ...  shuttle           11   20-03-18 5:00
5226    12544    22-03-18        7:04  ...      Bus           49   22-03-18 7:04
5390    12744    27-03-18        7:13  ...      Bus           49   27-03-18 7:13
860      7099    04-03-18        6:20  ...  shuttle           11   04-03-18 6:20

[4186 rows x 7 columns]

In [129]:
X_test.head()

ride_id travel_date travel_time  ... car_type max_capacity            date
2164     8771    04-02-18       10:53  ...  shuttle           11  04-02-18 10:53
611      6794    09-03-18        9:00  ...  shuttle           11   09-03-18 9:00
3524    10553    14-02-18       10:20  ...  shuttle           11  14-02-18 10:20
2292     8946    01-02-18       10:10  ...  shuttle           11  01-02-18 10:10
5084    12375    16-03-18        6:59  ...      Bus           49   16-03-18 6:59

[5 rows x 7 columns]

In [130]:
def  time_features(df): 
    df["date"]=pd.to_datetime(df["date"])
    df["dayofweek"]=df["date"].dt.dayofweek
    df["dayofyear"]=df["date"].dt.dayofyear
    df["dayofmonth"]=df["date"].dt.day
    df["year_woy"]=df["date"].dt.year.astype(str)+df["date"].dt.weekofyear.astype(str)
    df["hour"]=df["date"].dt.hour
    df["minute"]=df["date"].dt.minute
    df["is_weekend"]=df["dayofweek"].apply( lambda x : 1 if x  in [5,6] else 0 )
    df["year"]=df["date"].dt.year
    df["quarter"]=df["date"].dt.quarter
    df["month"]=df["date"].dt.month
    return df

In [131]:
X_train=time_features(X_train)

In [132]:
X_train.head()

ride_id travel_date travel_time  ...  year quarter  month
4376    11559    12-12-17        6:00  ...  2017       4     12
690      6887    08-12-17        5:10  ...  2017       3      8
2653     9409    29-01-18        5:20  ...  2018       1      1
602      6784    09-03-18       10:10  ...  2018       3      9
5863    13338    11-04-18        8:20  ...  2018       4     11

[5 rows x 17 columns]

In [133]:

X_test = time_features(X_test)

In [134]:
X_test.head()

ride_id travel_date travel_time  ...  year quarter  month
2164     8771    04-02-18       10:53  ...  2018       2      4
611      6794    09-03-18        9:00  ...  2018       3      9
3524    10553    14-02-18       10:20  ...  2018       1      2
2292     8946    01-02-18       10:10  ...  2018       1      1
5084    12375    16-03-18        6:59  ...  2018       1      3

[5 rows x 17 columns]

In [137]:
X_train.describe()

ride_id  max_capacity  ...      quarter        month
count   4186.000000   4186.000000  ...  4186.000000  4186.000000
mean    9989.658385     30.590062  ...     2.215002     5.699236
std     2287.621341     18.993104  ...     1.261174     4.058778
min     1442.000000     11.000000  ...     1.000000     1.000000
25%     8041.250000     11.000000  ...     1.000000     2.000000
50%    10064.000000     49.000000  ...     2.000000     4.000000
75%    11935.750000     49.000000  ...     4.000000    10.000000
max    20116.000000     49.000000  ...     4.000000    12.000000

[8 rows x 11 columns]